### ISSUE TITLE
(How) should data types be serialized?

### ISSUE THREAD

**Issue by Whathecode** [cph-cachet/carp.core-kotlin#14]
Data types are the 'glue' between measure requests and device capabilities:
- Devices will define the available data streams (by data type).
- Measures specify the type of data they want to measure.

![entity_model](https://user-images.githubusercontent.com/778006/42216240-736d03a6-7ec1-11e8-94e6-ee2af7a1f917.png)

During study protocol creation this allows triggering protocol warnings/errors when measures are triggered to devices which do not support said measures. At startup of the study environment (e.g., smartphone) after loading the study protocol, a warning can be triggered when not all data streams to be measured have a suitable 'probe' registered to be used for data collection.

Both devices and measures thus define data types in the study protocol. However, (how) should these be serialized?

- In the case of device data streams, we could simply hard code them so they are available if the corresponding device classes are loaded. The only down side is we cannot interpret data streams of unknown devices (specified in extending libraries) on the server side. Is there a use case where we have to?
- In essence, we only really need the type name in order to 'match' device data streams with measures. Any additional data in data types is likely only needed once we want to start specifying 'secondary' data streams which process other data streams on the device, e.g., higher-order interpretations of data. If we want to write triggers which are reusable for 'custom' data streams they will need 'some' info on how to interpret said data; this was the initial purpose of a 'data type' definition.

**Comment by Whathecode** [comment]
After picking up on this discussion, we have decided that, although there is value in having 'data type hierarchies', rather than a simple data type class with a 'namespace' and 'name' field, it is premature to try to design such a hierarchy at this point in time.

The following potential benefits still hold:

> Any additional data in data types is likely only needed once we want to start specifying 'secondary' data streams which process other data streams on the device, e.g., higher-order interpretations of data. If we want to write triggers which are reusable for 'custom' data streams they will need 'some' info on how to interpret said data; this was the initial purpose of a 'data type' definition.

_But_, it is hard to imagine exactly how said 'generic' processing of data would look like at this point in time (thus risking implementing complex data type hierarchies which turn out to be useless in the future), and, it is hard to assess whether alternate implementations would be just as useful in the future. For example, by having a registry of 'data parsers' which know how to extract/prepare data to common formats to be consumed by generic components in the infrastructure.

Thus, in conclusion, for now, we have decided to remove data type hierarchies and only use a single `DataType` class with the sole purpose of uniquely identifying the data (how it should be uploaded, what data can be measured). This also answers the question 'how should these be serialized', as all that remains is the identifier.
